proxy-server/background/proxy-handler.js

```diff
@@ -0,0 +1,114 @@
+
+      // Register the proxy script. Path is relative to manifest.json
+      browser.proxy.register(proxyScriptURL);
+      proxyScriptLoaded = true;
+      sendSettingsToProxyScript(proxyServerInfoAndAuth.proxyServerInfo);
+    }
+  }
+}
+
+function unregisterProxyScript() {
+  if (proxyScriptLoaded) {
+    console.log("Unegistering proxy script");
+    browser.proxy.unregister();
+    proxyScriptLoaded = false;
+  }
+}
+
+function validateProxyServerInfo(proxyServerInfo) {
+  if (proxyServerInfo && proxyServerInfo.proxyServerAddress &&
+        proxyServerInfo.proxyServerPort && proxyServerInfo.proxyServerType) {
+    console.log("Validated proxy server, proxy server port, and proxy type");
+    return true;
+  }
+  else {
+    console.log("Visit the options page to define a proxy server, proxy server port, and proxy type");
+    return false;
+  }
+}
+
+function sendSettingsToProxyScript(proxyServerInfo) {
+  console.log("Sending updated proxyServerInfo to proxy script");
+  browser.runtime.sendMessage(proxyServerInfo, {toProxyScript: true});
+}
+
+browser.runtime.onMessage.addListener(handleMessage);
+
+/**
+ * After https://bugzilla.mozilla.org/show_bug.cgi?id=1359693 is fixed, onAuthRequired() not needed.
+ * You can then specify username/password in FindProxyForURL().
+ * If incorrect username/password is specified, then Firefox displays either
+ * "Proxy server is refusing connections" (HTTP and HTTPS proxy servers) or
+ * "Unable to connect" (SOCKS proxy servers)
+ */
+browser.webRequest.onAuthRequired.addListener(provideCredentialsAsync,
+                                          {urls: ["<all_urls>"]},
+                                          ["blocking"]);
+
+
+// update the proxy whenever stored settings change
+browser.storage.onChanged.addListener((newProxyServerInfoAndAuth) => {
+  /** newProxyServerInfoAndAuth has this format (data is fake):
+   *
+   * proxyServerInfo: {newValue: {proxyServerAddress: "foo.bar", proxyServerPort: "3128", proxyServerType: "HTTP"},
+   *                  oldValue: {proxyServerAddress: "baz.bar", proxyServerPort: "1235", proxyServerType: "SOCKS"}},
+   * authCrendentials: {newValue: {username: "mungsantamaria", password: "mypass"}, oldValue: {username: "", password: ""}}
+   */
+  if (newProxyServerInfoAndAuth && validateProxyServerInfo(newProxyServerInfoAndAuth.proxyServerInfo.newValue)) {
+    if (proxyScriptLoaded) {
+      sendSettingsToProxyScript(newProxyServerInfoAndAuth.proxyServerInfo.newValue);
+    }
+    else {
+      registerProxyScript({proxyServerInfo: newProxyServerInfoAndAuth.proxyServerInfo.newValue,
+        authCredentials: newProxyServerInfoAndAuth.authCredentials.newValue});
+    }
+  }
+  else {
+    unregisterProxyScript();
+  }
+});
+
+// get the current settings, then...
+browser.storage.local.get()
+  .then((proxyServerInfoAndAuth) => {
+    // attempt to register the proxy script
+    registerProxyScript(proxyServerInfoAndAuth);
+  })
+  .catch(()=> {
```

Comment:
The old code distinguished between an error getting the settings, and successfully getting empty settings. This enabled it to log a helpful message on install:

```
  // get the current settings, then...
  browser.storage.local.get()
    .then((storedSettings) => {
      // if there are stored settings, update the proxy with them...
...
      }
      else {
        console.log("No proxy server defined: visit the options page to define a proxy server");
      }
    })
    .catch(()=> {
      console.error("Error retrieving stored settings");
    });
```
The new code doesn't do this, so you just see a generic storage error on startup.